File: graph/Leetcode-77-Combinations/Leetcode-77-Combinations.cpp
class Solution {
public:
    int n, k;
    
    void backtrack(int start, vector<int>& comb, vector<vector<int>>& combs){
        if(comb.size() == k){
            combs.push_back(comb);
        }else{
            for(int i = start; i <= n; ++i){
                comb.push_back(i);
                backtrack(i+1, comb, combs);
                comb.pop_back();
            }
        }
    }
    
    vector<vector<int>> combine(int n, int k) {
        this->n = n;
        this->k = k;
        
        vector<int> comb;
        vector<vector<int>> combs;
        
        backtrack(1, comb, combs);
        
        return combs;
    }
};

File: maureen72/Graph/Leetcode-77-Combinations.py
class Solution(object):
    def combine(self, n, k):
        results = []
        def dfs(comb, start):
            #base case
            if len(comb) == k:
                results.append(comb[:])
                return

            for i in range(start, n+1):
                comb.append(i)
                dfs(comb,i+1)
                comb.pop()
            
        dfs([],1)
        return results

        #return list(itertools.combinations(range(1,n+1), k))
        